From: Jonas Karlman <jonas@kwiboo.se>
Date: Sat, 4 Nov 2023 21:14:21 +0000
Subject: [09/11] avutil/pixfmt: Add NV15, NV20BS and NV30
Origin: https://github.com/Kwiboo/FFmpeg/commit/2fef1a1f108f7689d95a8705bfbdd349d983d273

Add NV15 and NV20 pixel formats used by the Rockchip Video Decoder for
10-bit buffers.

NV15 and NV20 is 10-bit 4:2:0/4:2:2 semi-planar YUV formats similar to
NV12 and NV16, using 10-bit components with no padding between each
component. Instead, a group of 4 luminance/chrominance samples are
stored over 5 bytes in little endian order:

YYYY = UVUV = 4 * 10 bits = 40 bits = 5 bytes

The '15' and '20' suffix refers to the optimum effective bits per pixel
which is achieved when the total number of luminance samples is a
multiple of 8 for NV15 and 4 for NV20.

The existing NV20LE/BE pixdesc indicate that the existing format is not
packed and instead have more in common with P210, each 10-bit sample is
LSB aligned instead of MSB aligned.

Instead a new similar name, NV20BS (bitstream), was chosen.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 libavutil/pixdesc.c              | 36 ++++++++++++++++++++++++++++++++
 libavutil/pixfmt.h               |  4 ++++
 tests/ref/fate/imgutils          |  6 ++++++
 tests/ref/fate/sws-pixdesc-query | 15 +++++++++++++
 4 files changed, 61 insertions(+)

diff --git a/libavutil/pixdesc.c b/libavutil/pixdesc.c
index 1c0bcf2232..e20ed725f6 100644
--- a/libavutil/pixdesc.c
+++ b/libavutil/pixdesc.c
@@ -578,6 +578,18 @@ static const AVPixFmtDescriptor av_pix_fmt_descriptors[AV_PIX_FMT_NB] = {
         },
         .flags = AV_PIX_FMT_FLAG_PLANAR,
     },
+    [AV_PIX_FMT_NV15] = {
+        .name = "nv15",
+        .nb_components = 3,
+        .log2_chroma_w = 1,
+        .log2_chroma_h = 1,
+        .comp = {
+            { 0, 10,  0, 0, 10 },      /* Y */
+            { 1, 20,  0, 0, 10 },      /* U */
+            { 1, 20, 10, 0, 10 },      /* V */
+        },
+        .flags = AV_PIX_FMT_FLAG_BITSTREAM | AV_PIX_FMT_FLAG_PLANAR,
+    },
     [AV_PIX_FMT_ARGB] = {
         .name = "argb",
         .nb_components = 4,
@@ -2079,6 +2091,18 @@ static const AVPixFmtDescriptor av_pix_fmt_descriptors[AV_PIX_FMT_NB] = {
         },
         .flags = AV_PIX_FMT_FLAG_PLANAR,
     },
+    [AV_PIX_FMT_NV20BS] = {
+        .name = "nv20",
+        .nb_components = 3,
+        .log2_chroma_w = 1,
+        .log2_chroma_h = 0,
+        .comp = {
+            { 0, 10,  0, 0, 10 },      /* Y */
+            { 1, 20,  0, 0, 10 },      /* U */
+            { 1, 20, 10, 0, 10 },      /* V */
+        },
+        .flags = AV_PIX_FMT_FLAG_BITSTREAM | AV_PIX_FMT_FLAG_PLANAR,
+    },
     [AV_PIX_FMT_NV20LE] = {
         .name = "nv20le",
         .nb_components = 3,
@@ -2103,6 +2127,18 @@ static const AVPixFmtDescriptor av_pix_fmt_descriptors[AV_PIX_FMT_NB] = {
         },
         .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BE,
     },
+    [AV_PIX_FMT_NV30] = {
+        .name = "nv30",
+        .nb_components = 3,
+        .log2_chroma_w = 0,
+        .log2_chroma_h = 0,
+        .comp = {
+            { 0, 10,  0, 0, 10 },      /* Y */
+            { 1, 20,  0, 0, 10 },      /* U */
+            { 1, 20, 10, 0, 10 },      /* V */
+        },
+        .flags = AV_PIX_FMT_FLAG_BITSTREAM | AV_PIX_FMT_FLAG_PLANAR,
+    },
     [AV_PIX_FMT_QSV] = {
         .name = "qsv",
         .flags = AV_PIX_FMT_FLAG_HWACCEL,
diff --git a/libavutil/pixfmt.h b/libavutil/pixfmt.h
index a7f50e1690..27649cb3b0 100644
--- a/libavutil/pixfmt.h
+++ b/libavutil/pixfmt.h
@@ -439,6 +439,10 @@ enum AVPixelFormat {
      */
     AV_PIX_FMT_D3D12,
 
+    AV_PIX_FMT_NV15,        ///< packed planar YUV 4:2:0, 15bpp, 1 plane for Y and 1 plane for the UV components, which are interleaved (first component U and the following component V), no padding between components
+    AV_PIX_FMT_NV20BS,      ///< packed planar YUV 4:2:2, 20bpp, 1 plane for Y and 1 plane for the UV components, which are interleaved (first component U and the following component V), no padding between components
+    AV_PIX_FMT_NV30,        ///< packed planar YUV 4:4:4, 30bpp, 1 plane for Y and 1 plane for the UV components, which are interleaved (first component U and the following component V), no padding between components
+
     AV_PIX_FMT_NB         ///< number of pixel formats, DO NOT USE THIS if you want to link with shared libav* because the number of formats might differ between versions
 };
 
diff --git a/tests/ref/fate/imgutils b/tests/ref/fate/imgutils
index fb2ed6d158..c8bc148f73 100644
--- a/tests/ref/fate/imgutils
+++ b/tests/ref/fate/imgutils
@@ -269,6 +269,9 @@ p412be          planes: 2, linesizes: 128 256   0   0, plane_sizes:  6144 12288
 p412le          planes: 2, linesizes: 128 256   0   0, plane_sizes:  6144 12288     0     0, plane_offsets:  6144     0     0, total_size: 18432
 gbrap14be       planes: 4, linesizes: 128 128 128 128, plane_sizes:  6144  6144  6144  6144, plane_offsets:  6144  6144  6144, total_size: 24576
 gbrap14le       planes: 4, linesizes: 128 128 128 128, plane_sizes:  6144  6144  6144  6144, plane_offsets:  6144  6144  6144, total_size: 24576
+nv15            planes: 2, linesizes:  80  80   0   0, plane_sizes:  3840  1920     0     0, plane_offsets:  3840     0     0, total_size: 5760
+nv20            planes: 2, linesizes:  80  80   0   0, plane_sizes:  3840  3840     0     0, plane_offsets:  3840     0     0, total_size: 7680
+nv30            planes: 2, linesizes:  80 160   0   0, plane_sizes:  3840  7680     0     0, plane_offsets:  3840     0     0, total_size: 11520
 
 image_fill_black tests
 yuv420p         total_size:   4608,  black_unknown_crc: 0xd00f6cc6,  black_tv_crc: 0xd00f6cc6,  black_pc_crc: 0x234969af
@@ -485,3 +488,6 @@ p412be          total_size:  18432,  black_unknown_crc: 0x26991800,  black_tv_cr
 p412le          total_size:  18432,  black_unknown_crc: 0x4028ac30,  black_tv_crc: 0x4028ac30,  black_pc_crc: 0xab7c7698
 gbrap14be       total_size:  24576,  black_unknown_crc: 0x4ec0d987,  black_tv_crc: 0x4ec0d987,  black_pc_crc: 0x4ec0d987
 gbrap14le       total_size:  24576,  black_unknown_crc: 0x13bde353,  black_tv_crc: 0x13bde353,  black_pc_crc: 0x13bde353
+nv15            total_size:   5760,  black_unknown_crc: 0x6b5fdb58,  black_tv_crc: 0x6b5fdb58,  black_pc_crc: 0x660a512c
+nv20            total_size:   7680,  black_unknown_crc: 0x171f53da,  black_tv_crc: 0x171f53da,  black_pc_crc: 0xfcf5cda3
+nv30            total_size:  11520,  black_unknown_crc: 0x664d0699,  black_tv_crc: 0x664d0699,  black_pc_crc: 0x4ab5d724
diff --git a/tests/ref/fate/sws-pixdesc-query b/tests/ref/fate/sws-pixdesc-query
index fff93bbf0e..9d2cf5e046 100644
--- a/tests/ref/fate/sws-pixdesc-query
+++ b/tests/ref/fate/sws-pixdesc-query
@@ -61,8 +61,11 @@ isNBPS:
   gray14le
   gray9be
   gray9le
+  nv15
+  nv20
   nv20be
   nv20le
+  nv30
   p010be
   p010le
   p012be
@@ -221,11 +224,14 @@ isYUV:
   ayuv64be
   ayuv64le
   nv12
+  nv15
   nv16
+  nv20
   nv20be
   nv20le
   nv21
   nv24
+  nv30
   nv42
   p010be
   p010le
@@ -337,11 +343,14 @@ isYUV:
 
 isPlanarYUV:
   nv12
+  nv15
   nv16
+  nv20
   nv20be
   nv20le
   nv21
   nv24
+  nv30
   nv42
   p010be
   p010le
@@ -434,11 +443,14 @@ isPlanarYUV:
 
 isSemiPlanarYUV:
   nv12
+  nv15
   nv16
+  nv20
   nv20be
   nv20le
   nv21
   nv24
+  nv30
   nv42
   p010be
   p010le
@@ -868,11 +880,14 @@ Planar:
   gbrpf32be
   gbrpf32le
   nv12
+  nv15
   nv16
+  nv20
   nv20be
   nv20le
   nv21
   nv24
+  nv30
   nv42
   p010be
   p010le
-- 
2.45.2

